File: 200. Number of Islands/NumberofIslands.py
class Solution(object):
    def numIslands(self, grid):
        """
        :type grid: List[List[str]]
        :rtype: int
        """
        if not grid or not grid[0]:
            return 0
        self.m=len(grid)
        self.n=len(grid[0])
        ans=0
        for i in range(self.m):
            for j in range(self.n):
                if grid[i][j]=='1':
                    ans+=1
                    self.bfs(i,j,grid)
        return ans
    def bfs(self,i,j,grid):
        grid[i][j]='0'
        if(i+1<self.m and grid[i+1][j]=='1'):
            self.bfs(i+1,j,grid)
        if(i-1>=0 and grid[i-1][j]=='1'):
            self.bfs(i-1,j,grid)
        if(j+1<self.n and grid[i][j+1]=='1'):
            self.bfs(i,j+1,grid)
        if(j-1>=0 and grid[i][j-1]=='1'):
            self.bfs(i,j-1,grid)

File: 200. Number of Islands/NumberofIslands.java
public class Solution {
    private int res;
    public int numIslands(char[][] grid) {
        //corner case.
        if(grid == null || grid.length == 0 || grid[0] == null || grid[0].length == 0)
            return 0;
        //a visited grid which has the same size as the input, indicates
        //if we had visited the according grid.
        boolean[][] visited = new boolean[grid.length][grid[0].length];
        //resolution.
        res = 0;
        //check on each individual island.
        for(int i = 0; i < grid.length; i++)
        {
            for(int j = 0; j < grid[0].length; j++)
            {
                if(visited[i][j] == false)
                {
                    if(helper(grid, visited, i, j))
                        res ++;
                }
            }
        }
        return res;
    }
    //it returns if the dfs result of the current location should be counted.
    public boolean helper(char[][] grid, boolean[][] visited, int currentRow, int currentCol )
    {
        //if the current location has been visited or it is sea.
        //return false to indicate that we should not take it into consideration.
        if(visited[currentRow][currentCol] == true || grid[currentRow][currentCol] == '0')
            return false;
        //else assign visited true.
        visited[currentRow][currentCol] = true;
        //do dfs flooding, so that each neighbor will be marked as visited.
        if(currentRow + 1 < visited.length)
        {
            helper(grid, visited, currentRow + 1, currentCol);
        }
        if(currentRow - 1 >= 0)
        {
            helper(grid, visited, currentRow - 1, currentCol);
        }
        if(currentCol + 1 < visited[0].length)
        {
            helper(grid, visited, currentRow, currentCol + 1);
        }
        if(currentCol - 1 >= 0)
        {
            helper(grid, visited, currentRow, currentCol - 1);
        }
        //return true to indicate that we find island!
        return true;
    }
}